lib/image_processing/vips.rb

```diff
@@ -0,0 +1,190 @@
+    #
+    # The resulting image will always be exactly as large as the specified
+    # dimensions.
+    #
+    # By default, the image will be placed in the center but this can be
+    # changed via the `gravity` option.
+    #
+    # @param [File, Tempfile] file      the image to convert
+    # @param [#to_s] width              the width to fill out
+    # @param [#to_s] height             the height to fill out
+    # @param [String] background        the color to use as a background
+    # @param [String] gravity           which part of the image to focus on
+    # @param [Path] destination_path    the end destination where the image will be safe
+    # @return [File, Tempfile]
+    # @see http://www.imagemagick.org/script/color.php
+    # @see http://www.imagemagick.org/script/command-line-options.php#gravity
+    def resize_and_pad(file, width, height, background: 'opaque', gravity: 'Center', destination_path: nil, &block)
+      with_ruby_vips(file, destination_path) do |vips_image|
+        vips_image = Utils.resize_image vips_image, width, height
+        top, left = Gravity.get(vips_image, width, height, gravity)
+        vips_image = vips_image.embed(top, left, width, height, {extend: :background, background: Color.get(background)})
+        vips_image
+      end
+    end
+
+    def resize_and_pad!(file, width, height, background: "opaque", gravity: "Center", &block)
+      resize_and_pad(file, width, height, background: background, gravity: gravity, destination_path: file.path, &block)
+    end
+
+    # Crops the image to be the defined area.
+    #
+    # @param [File, Tempfile] file      the image to convert
+    # @param [#to_s] width              the width of the cropped image
+    # @param [#to_s] height             the height of the cropped image
+    # @param [#to_s] x_offset           the x coordinate where to start cropping
+    # @param [#to_s] y_offset           the y coordinate where to start cropping
+    # @param [String] gravity           which part of the image to focus on
+    # @param [Path] destination_path    the end destination where the image will be safe
+    # @return [File, Tempfile]
+    # @see http://www.imagemagick.org/script/command-line-options.php#gravity
+    # @see http://www.vips.ecs.soton.ac.uk/supported/7.42/doc/html/libvips/libvips-conversion.html#vips-crop
+    def crop(file, width, height, gravity: 'NorthWest', destination_path: nil, &block)
+      with_ruby_vips(file, destination_path) do |vips_image|
+        top, left = Gravity.get(vips_image, width, height, gravity)
+        vips_image.crop top, left, width, height
+      end
+    end
+
+    def crop!(file, width, height, gravity: "NorthWest", &block)
+      crop(file, width, height, gravity: gravity, destination_path: file.path, &block)
+    end
+
+    # Convert an image into a Vips::Image for the duration of the block,
+    # and at the end return a File object.
+    def with_ruby_vips(file, destination_path, extension: nil)
```

Comment:
The `destination_path` should be an optional argument, because people might want to use this method directly, and it's better that instead of having to pass `nil` they can just not pass anything 😃